### PR TITLE
Add filter by skill level

### PR DESCRIPTION
**Pull request by jesse-spevack**
@rsbarbo and I added a section to the end of the markdown tutorial. We describe the steps we took to add filter functionality to the creact app. Our tutorial addition describes:

1. Creating a FilterSelect component
2. Wiring the FilterSelect component to a filter function in the Body component
3. Writing the filter functionality, which will set the state of the Body component and render all the filtered skills.
4. How to refactor the code to make it much clearer / DRY-er

@applegrain Rapha and I got a lot out of this tutorial and are glad we got a chance to contribute. Thanks for putting it together! You rock.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1598,10 +1598,1953 @@ end
 
 <br>
 
-### 12. You are awesome
+### 12. Filter Skills by Level
+---
+
+Our next step is to filter skills based on the level. First let's create a FilterSelect component.
```
Comment (biglovisa) [inline README.md]: `SelectFilter` might be a more idiomatic name

File: README.md [hunk 2]
```diff
@@ -1598,10 +1598,1953 @@ end
...
+
+```
+touch app/javascripts/components/_filter_selects.js.jsx
```
Comment (biglovisa) [inline README.md]: Usually file names are named the same as the component - so this should be `_filter_select.js.jsx` (or `_select_filter.js.jsx`

File: README.md [hunk 3]
```diff
@@ -1598,10 +1598,1953 @@ end
...
+```
+In our _filter_selects we will create a component that has a selector and label. It is important to highlight that the syntax below uses ES6 class declaration but the function is still the same as the previous method for declaring components.
+Our FilterSelect class renders a Label, and a Select input with 4 (four) options (all, bad, halfbad, and fantastic).
+
+```
+class FilterSelect extends React.Component {
```
Comment (biglovisa) [inline README.md:1615]: I'd prefer that we use the same syntax in the tutorial to avoid confusion. I don't mind using the ES6 class declaration, but we would just have to change the other components. 

Updating the syntax to ES6 might be a separate PR by itself.

File: README.md [hunk 4]
```diff
@@ -1598,10 +1598,1953 @@ end
...
+        <NewSkill handleSubmit={this.handleSubmit} />
+        <FilterSelect/>
+        <AllSkills skills={this.state.skills}
+                   handleDelete={this.handleDelete}
+                   handleUpdate={this.handleUpdate} />
+      </div>
+    )
+  }
+```
+If everything is wired correctly, you should now see a select input with the four options in between the create skill form and the AllSkills list.
```
Comment (biglovisa) [inline README.md:1650]: should it be: `... wired up correctly,`. 

Also, since you are calling it `AllSkills list`, I would also say `NewSkill form` to be consistent.

File: README.md [hunk 5]
```diff
@@ -1598,10 +1598,1953 @@ end
...
+
+So now we will actually create our filtering functionality. The big picture is that we want to filter based on the level, so will start our journey by passing a filter property to our FilterSelect component.
+
+_body.js.jsx
+```
+  render() {
+    return (
+      <div className="container">
+        <NewSkill handleSubmit={this.handleSubmit} />
+        <FilterSelect filter={this.filterSkills} />
```
Comment (biglovisa) [inline README.md:1661]: `this.filterSkills` is a function, so the prop passed down should be called `handleFilter` instead of just `filter` (it could read as a noun instead of an action).

File: README.md [hunk 6]
```diff
@@ -1598,10 +1598,1953 @@ end
...
+                   handleDelete={this.handleDelete}
+                   handleUpdate={this.handleUpdate} />
+      </div>
+    )
+  }
+```
+If everything is wired correctly, you should now see a select input with the four options in between the create skill form and the AllSkills list.
+<br>
+
+So now we will actually create our filtering functionality. The big picture is that we want to filter based on the level, so will start our journey by passing a filter property to our FilterSelect component.
```
Comment (biglovisa) [inline README.md:1653]: Here, I would specify "passing a handleFilter property to our FilterSelect component. This property we will call whenever we want to actually filter our skills, so we need to assign it a callback.". 

Then, you can simplify what you have on line 1669. for example: "We need to implement the callback we passed as a prop to the `SelectFilter` component in `_body.js.jsx`."

File: README.md [hunk 7]
```diff
@@ -1598,10 +1598,1953 @@ end
...
+  }
+```
+<br>
+
+I am sure you noticed that inside of our filter property we have a function called filterSkills, so let's go ahead and create that within our _body.js.jsx.
+
+```
+  filterSkills(value) {
+    console.log("about to filter");
+    debugger;
```
Comment (biglovisa) [inline README.md:1676]: remove the debugger

File: README.md [hunk 8]
```diff
@@ -1598,10 +1598,1953 @@ end
...
+  },
+```
+We will add the functionality of filtering in the above filterSkills(value) function. But first, let's wire it all together and make sure that when the user changes the select dropdown, this filterSkills function is invoked.
+<br>
+
+Let's add an onChange event to our select input that will call the updateFilter function. In order to accomplish that we must bind the object 'this' to our function call.
```
Comment (biglovisa) [inline README.md:1681]: Can you add a small explanation to why we need to use `bind` here? I don't remember if I have done it up top already.

File: README.md [hunk 9]
```diff
@@ -1598,10 +1598,1953 @@ end
...
+      <div>
+        <label>
+          Filter By
+        </label>
+        <select className="form-control" onChange={this.updateFilter.bind(this)}>
+        
+        ...
+```
+<br>
+Now we need to create an updateFilter function within the FilterSelect component.
```
Comment (biglovisa) [inline README.md:1696]: `the FilterSelect component which will be invoked whenever there is a change on our `<select>` element`. 

(I would also add more of the html in the code above)

File: README.md [hunk 10]
```diff
@@ -1598,10 +1598,1953 @@ end
...
+
+_filter_selects.js.jsx
+```
+updateFilter(event) {
+    this.props.filter(event.target.value)
+  }
+```
+Since we bound the object 'this' to the function call in our updateFilter, 'this' refers to the entire component. This gives us access to all of the properties of the FilterSelect component. We need the filter property, which we pass into the component when we render it in the Body. The filter property points to the filterSkills function in the Body component. Let's pass in the value of the select input, which we get from the event object.
```
Comment (biglovisa) [inline README.md]: This might be enough explanation of `bind` - above where I left the comment about `bind`, maybe just mention that you'll explain further down why we are `bind`ing.

Comment (biglovisa) [reply README.md]: Also, since there's a lot of describing of the events, maybe you could add a flow diagram? I think I have something similar up top.

File: README.md [hunk 11]
```diff
@@ -1598,10 +1598,1953 @@ end
...
+```
```
Comment (biglovisa) [inline README.md:1705]: This code should be
```
updateFilter(event) {
   this.props.filter(event.target.value)
}
```

And the one on line 1699 you could just `console.log(event.target.value)` instead.

File: README.md [hunk 12]
```diff
@@ -1598,10 +1598,1953 @@ end
...
+event.target.value
+```
+
+If everything is working properly, when you reload the page and select an option from the selector, you should hit the debugger within the filterSkills function. You should also see "about to filter" logged to the console. From the console, you can type "value" and you will get the value you selected from the dropdown.
```
Comment (biglovisa) [inline README.md:1709]: If you're referencing the debugger here, either specify to add it in this block - or add a comment up top that you have the debugger statement there so that you will get caught in it when it's wired up correctly.

File: README.md [hunk 13]
```diff
@@ -1598,10 +1598,1953 @@ end
...
+<br>
+Next we need to actually make the filterSkills function filter our skills list based on the value passed in from the select input.
+
+_body.js.jsx
+```
+filterSkills(value) {
+    // Create a variable called skills set to an empty string
+    var skills = ''
```
Comment (biglovisa) [inline README.md:1717]: use `let`

File: README.md [hunk 14]
```diff
@@ -1598,10 +1598,1953 @@ end
...
+event.target.value
+```
+
+If everything is working properly, when you reload the page and select an option from the selector, you should hit the debugger within the filterSkills function. You should also see "about to filter" logged to the console. From the console, you can type "value" and you will get the value you selected from the dropdown.
+<br>
+Next we need to actually make the filterSkills function filter our skills list based on the value passed in from the select input.
+
+_body.js.jsx
+```
+filterSkills(value) {
```
Comment (biglovisa) [inline README.md:1715]: `value` should be `filter` imo, you're passing it something to filter the list by.

File: README.md [hunk 15]
```diff
@@ -1598,10 +1598,1953 @@ end
...
+      // We can just set the state of the skills to be our skills variable
+      this.setState({ skills: skills });
+    } else {
+      // We additionally set allSkills to be the original list of skills.
+      this.setState({ skills: skills, allSkills: this.state.skills });
+    }
+  }
+```
+The above code will filter on 'bad', 'halfbad', and 'fantastic' skill levels, but has two big problems. First, the 'All' option will not return all of the skills. Seco# Creact
+## Crud + React
```
Comment (biglovisa) [inline README.md:1741]: It seems like there's some mixup with the code from here and down ....

File: README.md [hunk 16]
```diff
@@ -1598,10 +1598,1953 @@ end
...
+    if(this.state.allSkills) {
+      // We can just set the state of the skills to be our skills variable
+      this.setState({ skills: skills });
+    } else {
+      // We additionally set allSkills to be the original list of skills.
+      this.setState({ skills: skills, allSkills: this.state.skills });
+    }
+  }
+```
+The above code will filter on 'bad', 'halfbad', and 'fantastic' skill levels, but has two big problems. First, the 'All' option will not return all of the skills. Second, this code does not follow best practices because of the repetition and multiple conditional statements.
```
Comment (biglovisa) [inline README.md]: It seems like the code between this and the comment on line 1740 should go.

File: README.md (hunk 13 above)
Comment (biglovisa) [reply README.md:1717]: Also - this might just be me - but I find it incredibly difficult to read code that has explaining comments in it like below. I'd prefer if you explained and added code in a more iterative way (the way you would actually develop it) to make it easier to follow.

File: README.md [hunk 17]
```diff
@@ -1598,10 +1598,1953 @@ end
...
+event.target.value
+```
+
+If everything is working properly, when you reload the page and select an option from the selector, you should hit the debugger within the filterSkills function. You should also see "about to filter" logged to the console. From the console, you can type "value" and you will get the value you selected from the dropdown.
+<br>
+Next we need to actually make the filterSkills function filter our skills list based on the value passed in from the select input.
+
+_body.js.jsx
+```
+filterSkills(value) {
```
Comment (biglovisa) [inline README.md]: I'm not entirely sure I understand this function. Can we only filter by levels? If so, we should be explicit about that in the beginning of this section. 

And the implementation of this function could perhaps just be: 
```
filterSkills(filter) {
   const skills =  this.state.skills.filter(skill => skill.level === filter);
   this.setState({ skills });
}
```

You would just have to have a check that when there is no filter (when you `deselectFilter`, you do another API call and set `this.state.skills` to whatever the response is.

File: README.md [hunk 18]
```diff
@@ -1598,10 +1598,1953 @@ end
...
+The above code will filter on 'bad', 'halfbad', and 'fantastic' skill levels, but has two big problems. First, the 'All' option will not return all of the skills. Second, this code does not follow best practices because of the repetition and multiple conditional statements.
+
+It's time to refactor and figure out how to handle the 'all' functionality. Lets start by handling the 'all'. 
+The best way to do this, is to create a new property for our Body within our ComponentDidMount function. We will do this because we want to avoid making additional Ajax calls when the user wants to filter by different levels.
+
+_body.js.jsx
+```
+componentDidMount() {
+   $.getJSON('/api/v1/skills.json', (response) => {
+     this.setState({ skills: response,
```
Comment (biglovisa) [inline README.md]: See what I commented above, not the biggest fan of keeping two copies. It's cheap enough to go back to the server and ask for all skills.

**Review by biglovisa — COMMENTED** [review]
This looks really good @PlanetEfficacy @rsbarbo!! 

Please ask any questions you have on my comments, or correct me if I misunderstood something.

File: README.md (hunk 1 above)
Comment (jesse-spevack) [reply README.md]: Check!

File: README.md (hunk 2 above)
Comment (jesse-spevack) [reply README.md]: Check!

File: README.md (hunk 3 above)
Comment (jesse-spevack) [reply README.md:1615]: Check!

File: README.md (hunk 4 above)
Comment (jesse-spevack) [reply README.md:1650]: Check!

File: README.md (hunk 6 above)
Comment (jesse-spevack) [reply README.md:1653]: Check!

File: README.md (hunk 5 above)
Comment (jesse-spevack) [reply README.md:1661]: Check!

File: README.md (hunk 7 above)
Comment (jesse-spevack) [reply README.md:1676]: Check

File: README.md (hunk 8 above)
Comment (jesse-spevack) [reply README.md:1681]: check!

File: README.md (hunk 9 above)
Comment (jesse-spevack) [reply README.md:1696]: Check!

File: README.md (hunk 10 above)
Comment (jesse-spevack) [reply README.md]: Check!

File: README.md (hunk 12 above)
Comment (jesse-spevack) [reply README.md:1709]: Check!

File: README.md (hunk 14 above)
Comment (jesse-spevack) [reply README.md:1715]: check!

File: README.md (hunk 13 above)
Comment (jesse-spevack) [reply README.md:1717]: check!

File: README.md (hunk 14 above)
Comment (jesse-spevack) [reply README.md:1715]: I changed it to level, since I renamed the function `filterSkillsByLevel`

File: README.md (hunk 15 above)
Comment (jesse-spevack) [reply README.md:1741]: Check!

File: README.md (hunk 18 above)
Comment (jesse-spevack) [reply README.md]: As per your suggestion I've switched the implementation in the tutorial to do an API call instead of storing a second copy of the skills list in the app's state.

**Comment by jesse-spevack** [comment]
@applegrain Great feedback overall. You have high standards, which I respect. Please review my responses and changes to the markdown file. Hopefully these changes make the steps clearer. More feedback is always welcome.

**Comment by jesse-spevack** [comment]
Hey @applegrain I'm feeling good about `skills={this.state.filteredSkills || this.state.skills}`. Really nice solution. I updated section 12 of the tutorial to reflect this design change. Thanks again!

File: README.md (hunk 15 above)
Comment (biglovisa) [reply README.md:1741]: It seems like this is still happening, the `sections` piece is below, and then you have another copy of section 12 after that. 

The diff is a bit weird, now it looks like you are adding the entire tutorial. To get a cleaner diff, I would remove everything you have above, and update the section below.

File: README.md [hunk 19]
```diff
@@ -1598,10 +3338,203 @@ end
...
+
+**app/assets/components/javascripts/_body.js.jsx**
+```
+filterSkillsByLevel(level) {
+  console.log("about to filter");
+},
+```
+We will add the functionality of filtering in the above `filterSkillsByLevel(level)` function. But first, let's wire it all together and make sure that when the user changes the select dropdown, this `filterSkillsByLevel` function is invoked.
+
+
```
Comment (biglovisa) [inline README.md:3419]: extra newline

File: README.md [hunk 20]
```diff
@@ -1598,10 +3338,203 @@ end
...
+Let's add an `onChange` event to our select input that will call the `updateFilter` function. In order to accomplish that we must bind the object 'this' to our function call. In this case, the 'this' object refers to the `SelectFilter` component. This is important because if we did not bind 'this', when we get into the updateFilter function call we would not be able to call `this.state` or `this.props` since in that case, this would refer to the function `updateFilter`, not the component `SelectFilter`. That is a lot of this's and thats and for what its worth, this is probably the toughest part of React!
```
Comment (biglovisa) [inline README.md:3420]: great!! :tada:

File: README.md [hunk 21]
```diff
@@ -1598,10 +3338,203 @@ end
...
+Now we need to create an updateFilter function within the FilterSelect component which will be invoked whenever there is a change on our element.
+
+**app/assets/components/javascripts/_select_filter.js.jsx**
+```
+updateFilter(event) {
+  this.props.handleFilter(event.target.value)
+}
+```
+Remember, since we bound the object 'this' to the function call in our `updateFilter`, 'this' refers to the entire component. This gives us access to all of the properties of the `SelectFilter` component. We need the `handleFilter` property, which we pass into the component when we render it in the Body. The `handleFilter` property points to the `filterSkillsByLevel` function in the `Body` component. 
+```
```
Comment (biglovisa) [inline README.md:3443]: :tada:

File: README.md [hunk 22]
```diff
@@ -1598,10 +3338,203 @@ end
...
+         Main
+       /      \
+  Header        Body
+            /      \                 \ 
+        NewSkill   SelectFilter    AllSkills
+                                       \
+                                      Skills * n
+```
+For additioanl context, in the above diagram when the user selects a value to filter by in the `SelectFilter` component, the `handleFilter` property is invoked which sends a message up to the `body` component, which in turn re-renders the list of visible skills.
```
Comment (biglovisa) [inline README.md]: typo: `additional`

Maybe also specify that when we send a message up to the body component, we perform a state change which is passed down to SelectFilter, and the state change is what would cause the re-render.

File: README.md [hunk 23]
```diff
@@ -1598,10 +3338,203 @@ end
...
+
+Let's pass in the value of the select input from the `SelectFilter` to the `Body`, which we get from the event object.
+```
+event.target.value
+```
+If everything is working properly, when you reload the page and select an option from the selector, you should see "about to filter" logged to the console.
+
+
```
Comment (biglovisa) [inline README.md:3460]: extra newline

**Comment by biglovisa** [comment]
@PlanetEfficacy - this is great! The diff is still a bit weird, but the last Filter section is the one you should keep. 

nice work! :1st_place_medal:

File: README.md (hunk 19 above)
Comment (jesse-spevack) [reply README.md:3419]: Check!

File: README.md (hunk 22 above)
Comment (jesse-spevack) [reply README.md]: How about: 

For additional context, in the above diagram when the user selects a value to filter by in the `SelectFilter` component, the `handleFilter` property is invoked which sends a message up to the `body` component. This message then changes the state of our `body` component which in turn re-renders the list of visible skills.

File: README.md (hunk 23 above)
Comment (jesse-spevack) [reply README.md:3460]: Check!

**Comment by jesse-spevack** [comment]
Hey @applegrain I updated my PR based on the above feedback. I appreciate your exacting standards! Thanks!

**Comment by biglovisa** [comment]
nice work! @PlanetEfficacy @rsbarbo
